scylla.go

```diff
@@ -308,6 +352,25 @@ func (p *scyllaConnPicker) Put(conn *Conn) {
 	}
 }
 
+// Returns a shard to connect to. If there is a
+func (p *scyllaConnPicker) getShardToConnect() (int, bool) {
+	if p.nrConns == p.nrShards {
+		return 0, false
+	}
+	// There is at least one shard that does not have a connection open.
+	// Advance the counter in round robin fashion until the shard is found.
+	for {
```

Comment:
Opening a new TCP conn is a rare and costly event. For the sake of simplicity I'd suggest implementing it as a plain loop over `p.conns` where you return id of the first occurance of `p.conns[shard] == nil`

Reply:
Aren't we opening many connections soon after the session is created? AFAIK, for each host:

- We open the first connection and get information about shards,
- Then we open nshards - 1 connections in parallel.

We put a connection to `p.conns` only after the connection is established.
If we always start looking from the beginning of the slice, all nshards - 1 connections will try to connect to the same shard.

Reply:
We need to handle that.

Another idea could be changing the way `scyllaConnPicker` works so that it opens connections on it's own but that is an abuse IMO.

Reply:
In v2, I added an atomic round-robin counter which determines which shard we will connect to.

OTOH, if a situation happens in which the pool is full, the connection to shard X gets dropped, and the counter is set so that next attempt will be made to shard X+1, then we will open connections to all N-1 other shards before we get the right one.

Should we handle that as well? I'm not sure how to do this without a proper mutex and a loop over `p.conns` like in v1 of this PR.

Reply:
If we have a dedicated type separated from scyllaConnPicker it can have it's own mutex.
Then you can have a bitmap or other means to know what you have already tried with a dialer.